File: leetcode-solutions/ruby/83. Remove Duplicates from Sorted List.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {ListNode}
def delete_duplicates(head)
    current_node = head

    while(current_node && current_node.next)
        # remove next node if current val and next value is equal
        if current_node.val == current_node.next.val
            current_node.next = current_node.next.next
        else
            current_node = current_node.next  
        end
        
    end
    
    return head
end

# diff solution
def delete_duplicates(head)
    prev = nil
    cur = head
    while cur
        if prev && prev.val == cur.val
            # remove all the duplicate
            while cur && cur.val == prev.val
                cur = cur.next
            end
            prev.next = cur
            prev = cur
            # if cur is nil
            cur = cur.next if cur
        else
            prev = cur
            cur = cur.next
        end
    end
    
    head
end

File: leetcode-solutions/python/83. Remove Duplicates from Sorted List.py
# https://leetcode.com/problems/remove-duplicates-from-sorted-list/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None or head.next == None:
            return head
        
        prev = head
        current = head.next
        while current != None:
            if prev.val != current.val:
                prev.next = current
                prev = current
            
            current = current.next
        prev.next = None
        return head
